docs/proposals/03-networking.md

```diff
@@ -0,0 +1,72 @@
+# Network Extensibility
+
+Currently Gardener follows a mono network-plugin support model (i.e., Calico). Although this can seem to be the more stable approach, it does not completely reflect the real use-case. This proposal brings forth an effort to add an extra level of customizability to Gardener networking.
+
+## Motivation
+
+Gardener is an open-source project that provides a nested user model. Basically, there are two types of services provided by Gardener to its users:
+
+- **Managed**: users only request a Kubernetes cluster (Clusters-as-a-Service)
+- **Hosted**: users utilize Gardener to provide their own managed version of Kubernetes (Cluster-Provisioner-as-a-service)
+
+For the first set of users, the choice of network plugin might not be so important, however, for the second class of users (i.e., Hosted) it is important to be able to customize networking based on their needs.
+
+Furthermore, Gardener provisions clusters on different cloud-providers with different networking requirements. For example, Azure does not support Calico Networking [1], this leads to the introduction of manual exceptions in static add-on charts which is error prune and can lead to failures during upgrades.
+
+Finally, every provider is different, and thus the network always needs to adapt to the infrastructure needs to provider better performance. Consistency does not necessarily lie in the implementation but in the interface.
+
+## Gardener Network Extension
+
+The goal of the Gardener Network Extensions is to support different network plugin, therefore, the specification for the network resource won't be fixed and will be customized based on the underlying network plugin. To do so, a `NetworkConfig` field in the spec will be provided where each plugin will define. Below is an example for deploy Calico as the cluster network plugin.
+
+```yaml
+---
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Network
+metadata:
+  name: calico-network
+  namespace: shoot--core--test-01
+spec:
+  type: calico
```

Comment:
Not sure how this can be achieved - we already know that calico needs to be configured differently per cloud provider. This is not reflected here in this spec. Also, can you explain how this information (which network plugin should be used) flows from the `Shoot` to this resource?

Reply:
I don't think this spec should know about the provider, the creator of the spec knows which provider this network configuration is meant for and the spec should provide enough flexibility to be configured for different underlying providers (e.g., bird). Also, didn't include it in this PR, but the information should flow the same way the infrastructure info flows, basically something like: 

```yaml
apiVersion: gardener.cloud/v1alpha1
kind: Shoot
metadata:
  name: test-shoot
  namespace: garden-dev
spec:
  network:
    type: calico
    clusterCIDR: 1.2.3.4/24 
    serviceCIDR: 4.3.2.1/24 
    providerConfig:
      apiVersion: calico.cloud.gardener.cloud/v1alpha1
      kind: NetworkConfig
      ....
```

Reply:
That would mean that the end-user has to configure all that, requiring knowledge about the CNI plugin that most end-users don't have and don't care about. I agree that we allowing to configure it is great, however, we should find a way to make it easy for the 90% of the users that just want a cluster with calico, that's it. Do you have an idea how to achieve that?

Reply:
Optional parameter and Default? If the user doesn’t care just default to Calico. Otherwise, use the configuration provided. At least that’s how I imagined it.

Reply:
Optional parameter and Default? If the user doesn’t care just default to Calico. Otherwise, use the configuration provided. At least that’s how I imagined it.

Reply:
> Optional parameter and Defaulting?

That would require a webhook in the garden cluster which we don't have today.

>  you already have the cloud-provider information at this point.

Who has this information? Gardener cannot do anything, it can just transport information from the shoot resource to extension resources.

Generally, this network extension resource opens up another dimension. In the context of #308 we should probably start smaller and just allow one single network plugin per cloud provider. Otherwise, I'm not sure how all of this can be achieved short-term (and we need a solution within the next weeks to get rid of the major blockers for other providers to integrate). Not speaking of validating Gardener in combination with all the different plugins/configurations.

What do you think? Can we start with a much simpler approach that uses `.spec.type={aws,azure,...}` and the respective provider controller deploys one CNI plugin ("hard-coded" which one) with the proper configuration? With more experience we can maybe open this dimension up and support also other network plugins, however, then we have to think of a solutions for the above mentioned problems.

Reply:
The goal of this network extension is to open up the door for plugin choice but not necessarily doing it now. Still we need the design to be flexible, enforcing a mono-plugin approach with a flexible spec is fine but designing the spec to be cloud provider specific means that changing the spec in the future will be a breaking change. However, I do understand the concern,  we can provide both in the spec, namely spec.cloudprovider=<aws,gcp,azure,...> and spec.type=<calico,cilium,azurecni,...>. The type needs to reflect the resource, since this is not meant to be cloud-provider specific but rather a hack introducing an ephemeral field such as cloud-provider is easy to deal with and won't affect how the extension is designed. 

When we move to support a different plugin, we can simply drop or ignore the `spec.cloudprovider` field. Does this sound reasonable? 

Another approach is to read the cloud-provider information from the cluster resource, but I guess that is more work.

Reply:
Thanks @zanetworker, this sounds good! Personally, I would now try to keep the API minimal such that we can progress with moving the networking part out-of-tree to the extension controllers soon. Later, when we have more capacity to focus on more sophisticated configurability options for the CNI plugin we can open up the API with the needed fields.

First, we should not enforce the user to specify the network plugin in the `Shoot`. Gardener should write the `Network` extension resource to the seed, however, without `.spec.type` but just containing `.spec.cloudProvider`. Then, the provider extension controllers can act upon that and deploy the ONE implemented CNI plugin for their environment (AWS uses Calico, Azure might use Azure CNI, ...).

So, my proposal is to introduce

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: Network
metadata:
  name: calico-network
  namespace: shoot--core--test-01
spec:
  cloudProvider: {aws,azure,...}
status:
  observedGeneration: 2
  lastOperation: ...
  lastError: ...
```

and to not change the `Shoot` resource right now.
This should give us enough freedom to open and extend it in the future, and it does not complicate the implementation for our short-term goal (moving cloud specifics out of gardener) right now.

Reply:
Alright, as discussed, the provider spec info will be there but won't be used for the first implementation and we will keep the **spec.cloudprovider** and **spec.type**, along with the CIDRs and apply directly to the shoot.